scripts/compileRoomTypesTuple.ts

```diff
@@ -1,17 +1,42 @@
 import fs from 'fs';
-import { roomStateSchema } from '../src/utils/types/roomStateSchema';
+import { roomStateDefinitionNameValueTuples } from '../src/utils/types/roomStateSchema';
 
-const roomTypes: string[] = roomStateSchema.oneOf.map(
-  (schema) => schema.properties.type.const,
+const schemaTypeScriptTypes = roomStateDefinitionNameValueTuples.map(
+  ([schemaName, schema]) => ({
+    roomStateTypeIdentifier: schemaName,
+    roomTypeType: `'${schema.properties.type.const}'`,
+  }),
+);
+
+const typeScriptTypeImportList = schemaTypeScriptTypes
+  .map(({ roomStateTypeIdentifier }) => roomStateTypeIdentifier)
+  .join(', ');
+
+const roomTypeTypes: string[] = schemaTypeScriptTypes.map(
+  ({ roomTypeType }) => roomTypeType,
 );
 
 const output = [
   '// THIS FILE WAS AUTOMATICALLY GENERATED',
   '// Use "npm run compile:gameStateSchema" to rebuild',
   '',
+  '// eslint-disable-next-line import/no-restricted-paths',
+  `import { ${typeScriptTypeImportList} } from './gameState';`,
+  '',
   'export const ROOM_TYPES_TUPLE = [',
-  ...roomTypes.map((roomType) => `  '${roomType}',`),
+  ...roomTypeTypes.map((roomTypeType) => `  ${roomTypeType},`),
   '] as const;',
+  '',
+  'export type RoomTypesTuple = typeof ROOM_TYPES_TUPLE;',
+  '',
+  'export type RoomType = RoomTypesTuple[number];',
+  '',
+  'export type NarrowedRoomState<TRoomType extends RoomType> =',
+  ...schemaTypeScriptTypes.map(
+    ({ roomStateTypeIdentifier, roomTypeType }) =>
+      `  TRoomType extends ${roomTypeType} ? ${roomStateTypeIdentifier} :`,
+  ),
+  '  never;',
```

Comment:
What does `never` do in this context?

Reply:
It really doesn't do anything. I just thought it would be easier to read if every RoomType was explicitly mapped to a RoomState. Since these types are generated together all RoomTypes will be mapped to a RoomState. I have done some local testing where I manually added a RoomType to the roomTypes tuple, and so anything that tries to derive a RoomState from the made up RoomType will be assigned a type of `never`. Which means it can't be assigned data, and will throw a lot of type errors.